rollbar-java/src/main/java/com/rollbar/notifier/Rollbar.java

```diff
@@ -92,19 +92,37 @@ public void handleUncaughtErrors(Thread thread) {
    * Replace the configuration of this instance.
    * This {@link ConfigBuilder} passed to configProvider is
    * preconfigured with the values of the current configuration.
-   * This method potentially block to acquire a write lock when
-   * replacing the configuration.
+   * This method potentially blocks to acquire a locks when
+   * safely work with the configuration.
    *
    * @param configProvider the provider of a new configuration
    */
   public void configure(ConfigProvider configProvider) {
-    this.configWriteLock.lock();
+    ConfigBuilder builder;
+
+    this.configReadLock.lock();
     try {
-      ConfigBuilder builder = ConfigBuilder.withConfig(this.config);
-      this.config = configProvider.provide(builder);
+      builder = ConfigBuilder.withConfig(this.config);
     } finally {
-      this.configWriteLock.unlock();
+      this.configReadLock.unlock();
     }
+
+    Config newConfig = configProvider.provide(builder);
+
+    this.configWriteLock.lock();
```

Comment:
I think here we could use the `configure(Config config)` method. 
Also, I'd wrap all the operations after acquiring the lock in a `try` statement, and the release of them in the `finally` clause, although the code is more verbose is a kind of pattern that you always see and could avoid some unexpected behaviors if weird errors happen.